src/control/server/storage/bdev/firmware.go

```diff
@@ -147,7 +165,58 @@ func getDeviceController(pciAddr string, controllers storage.NvmeControllers) (*
 		}
 	}
 
-	return nil, fmt.Errorf("no NVMe controller found with PCI address %q", pciAddr)
+	return nil, FaultPCIAddrNotFound(pciAddr)
+}
+
+func filterControllers(controllers storage.NvmeControllers, modelID, fwRev string) storage.NvmeControllers {
+	selected := make(storage.NvmeControllers, 0, len(controllers))
+	for _, ctrlr := range controllers {
+		if filterMatches(modelID, ctrlr.Model) &&
+			filterMatches(fwRev, ctrlr.FwRev) {
+			selected = append(selected, ctrlr)
+		}
+	}
+	return selected
+}
+
+func filterMatches(filterStr, actualStr string) bool {
+	return filterStr == "" || strings.ToUpper(actualStr) == strings.ToUpper(filterStr)
+}
+
+// UpdateFirmware updates the NVMe device controller firmware.
+func (p *Provider) UpdateFirmware(req FirmwareUpdateRequest) (*FirmwareUpdateResponse, error) {
+	if p.shouldForward(req) {
+		return p.fwFwd.Update(req)
+	}
+
+	if len(req.FirmwarePath) == 0 {
+		return nil, errors.New("missing path to firmware file")
+	}
+
+	controllers, err := p.getRequestedControllers(req.DeviceAddrs, req.ModelID, req.FirmwareRev, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(controllers) == 0 {
+		return nil, errors.New("no NVMe device controllers")
```

Comment:
Rather than returning a simple error here, it might be better to make this a fault.Fault with a more verbose message and resolution, e.g. "no NVMe device controllers matched the filter criteria", "adjust or relax the filters and try again".

I think it's also important to distinguish between "zero controllers because there are no controllers in the system", and "zero controllers because the filters were too tight".

Reply:
Fixed. I differentiated between no devices at all vs. no matches to filter, too. I left the "no devices in system" case as an ordinary error, because I don't think there's really any resolution for that.